business/layer.go

```diff
@@ -18,33 +18,48 @@ type Layer struct {
 	OpenshiftOAuth OpenshiftOAuthService
 }
 
-// Global business.Layer; currently only used for tests to inject mocks,
-//	whereas production code recreates services in a stateless way
-var layer *Layer
+// Global clientfactory and prometheus clients.
+var clientFactory kubernetes.ClientFactory
+var prometheusClient prometheus.ClientInterface
 
-// Get the business.Layer, create it if necessary
-func Get() (*Layer, error) {
-	if layer != nil {
-		return layer, nil
+func GetUnauthenticated() (*Layer, error) {
+	return Get("")
+}
+
+// Get the business.Layer
+func Get(token string) (*Layer, error) {
+	// Use an existing client factory if it exists, otherwise create and use in the future
+	if clientFactory == nil {
+		userClient, err := kubernetes.NewClientFactory()
+		if err != nil {
+			return nil, err
+		}
+		clientFactory = userClient
 	}
-	k8s, err := kubernetes.NewClient()
+
+	// Creates a new k8s client based on the current users token
+	k8s, err := clientFactory.NewClient(token)
```

Comment:
I wonder if we can maintain a map with token -> client with some time to expire (configurable).
That would be easy to implement and it would help to reduce the client operation at cost to maintain some memory increasy.
It is expected that same token will be re-used multiple times so that probably could make sense.

Reply:
Yeah, we could do this.

I wonder if we want to do this purely based on expiration, or do something like store X many clients and when we go over X we kick out the one that was used last (so we known a bound on the memory size).

Reply:
I simple first version could be ok, with some configuration for some tweaks.
But I wouldn't add too much effort yet on it.
A map with last accessed and a clean/purge if last access was > threshold could work.
That could give us also hints if we can apply k8s caches on it or if a next effort is not worth.